### PR TITLE
Convert cyclic to a record; refactor code to reduce duplication with Block

### PR DESCRIPTION
**Pull request by bradcray**
This PR converts Cyclic to use a record, similar to what was done for Block in #22657.

To avoid code duplication, it refactors code that is common to both distributions (and the historical `_distribution` type) into a new `chpl_PrivatizedDistHelper` record type, defined in `DSIUtil.chpl`.  Then, each of `Block` and `Cyclic` store an instance of this record as a forwarding field.  This provides a subclassing-like (mix-in?) behavior for records.  As future work, I found myself what would be required to factor even more code into it (essentially all routines that currently refer to the `Block` or `Cyclic` types directly), but am not taking that on here in order to focus on removing the need for `dmap` for as many distributions as possible.

The steps I used to get here were:
* cloned `Block` into `Cyclic` and specialize it for Cyclic as necessary
* got `Cyclic` working as a standalone record
* copied common aspects of `Block` and `Cyclic` into a new `chpl_PrivatizedDistHelper` type
* updated tests and module code to avoid using `... dmap(... Cyclic)`
* updated my deprecation test to include Cyclic as well as Block

### CODE REVIEW

File: modules/dists/BlockDist.chpl [hunk 1]
```diff
@@ -348,157 +344,38 @@ record Block {
...
 
     proc init(value) {
       this.rank = value.rank;
       this.idxType = value.idxType;
       this.sparseLayoutType = value.sparseLayoutType;
-      this._pid = if _isPrivatized(value) then _newPrivatizedClass(value) else nullPid;
-      this._instance = _to_unmanaged(value);
+      this.chpl_distHelp = new chpl_PrivatizedDistHelper(
+                                                         if _isPrivatized(value) then _newPrivatizedClass(value) else nullPid,
+                                                         _to_unmanaged(value));
```
Comment (bmcdonald3) [inline modules/dists/BlockDist.chpl:367]: Nit, but it looks like this indentation is significantly different than that in the initializer in 347-349. I definitely prefer the syntax from 347-349 (and don't think this one fits within the character limit).

File: modules/dists/CyclicDist.chpl [hunk 2]
```diff
@@ -183,7 +183,105 @@ where the updates are made.
...
+                                                         _unowned);
+    }
+
+    proc init(value) {
+      this.rank = value.rank;
+      this.idxType = value.idxType;
+
+      this.chpl_distHelp = new chpl_PrivatizedDistHelper(
+                                                         if _isPrivatized(value) then _newPrivatizedClass(value) else nullPid,
+                                                         _to_unmanaged(value));
```
Comment (bmcdonald3) [inline modules/dists/CyclicDist.chpl:233]: Same comment as in Block, the indentation appears off here.

File: modules/dists/CyclicDist.chpl [hunk 3]
```diff
@@ -183,7 +183,105 @@ where the updates are made.
...
+operator =(ref a: Cyclic(?), b: Cyclic(?)) {
+  if a._value == nil {
+    __primitive("move", a, chpl__autoCopy(b.clone(), definedConst=false));
+  } else {
+    if a._value.type != b._value.type then
+      compilerError("type mismatch in distribution assignment");
+    if a._value == b._value {
+      // do nothing
+    } else
+        a._value.dsiAssign(b._value);
```
Comment (bmcdonald3) [inline modules/dists/CyclicDist.chpl:276]: Was this "do nothing" branch added to make things more readable? It seems just using `if a._value != b._value then a._value.dsiAssign(b._value);` or something like that would be simpler; this empty branch seems odd to me.

**Review by bmcdonald3 — APPROVED** [review]
Didn't have too much to say technically, just a few stylistic comments. Clean up looks good and happy to see us moving away from `dmap` 👍.

File: modules/dists/BlockDist.chpl (hunk 1 above)
Comment (bradcray) [reply modules/dists/BlockDist.chpl:367]: Good point, I'll change all of these to be more like:
```chapel
...new blah(
     if blah
       then blah
       else blah,
     nextArg);
```

File: modules/dists/CyclicDist.chpl (hunk 3 above)
Comment (bradcray) [reply modules/dists/CyclicDist.chpl:276]: I don't know, as this is just code that I've been copying over from the existing `_distribution` type, specializing it for Block and Cyclic without modifications other than the type signatures:  https://github.com/chapel-lang/chapel/blob/main/modules/internal/ChapelArray.chpl#L2300

Looking at the PR that added it, it appears that this branch used to contain some code and that Michael may have replaced the code with the comment to minimize the diff / retain the previous control flow structure:  https://github.com/chapel-lang/chapel/pull/4762/files#diff-52d2477eb48a33d9e81b64d64d74a50145a1bacdd880825e267b5bd5549559afR2964

**Pull request merged**